File: app/src/main/java/com/spiderbiggen/randomchampionselector/ddragon/DDragon.java
package com.spiderbiggen.randomchampionselector.ddragon;

import android.content.SharedPreferences;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.annotation.NonNull;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import com.spiderbiggen.randomchampionselector.R;
import com.spiderbiggen.randomchampionselector.model.Champion;
import com.spiderbiggen.randomchampionselector.model.ImageType;
import com.spiderbiggen.randomchampionselector.storage.file.FileStorage;
import com.spiderbiggen.randomchampionselector.util.async.ProgressCallback;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import okhttp3.ResponseBody;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.spiderbiggen.randomchampionselector.util.async.ProgressCallback.Progress.VERIFY_SUCCESS;

/**
 * Created on 13-3-2018.
 *
 * @author Stefan Breetveld
 */
public class DDragon {

    private static final String TAG = DDragon.class.getSimpleName();
    private static final DDragon instance = new DDragon();
    private static final String BASE_URL = "http://ddragon.leagueoflegends.com";
    private static final String DEFAULT_VERSION = "8.4.1"; // Default version if versions endpoint fails
    private static final AtomicReference<String> version = new AtomicReference<>(DEFAULT_VERSION);
    private static final int MAX_CONCURRENCY = 8;
    private final DDragonService service;
    private SharedPreferences preferences;
    private Resources resources;

    private DDragon() {
        this.service = createService();
    }

    public static DDragon getInstance() {
        return instance;
    }

    /**
     * Sets preferences.
     *
     * @param preferences the new value of preferences
     */
    public void setPreferences(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    /**
     * Sets resources.
     *
     * @param resources the new value of resources
     */
    public void setResources(Resources resources) {
        this.resources = resources;
    }

    public Disposable updateVersion(@NonNull Action onComplete, @NonNull Consumer<Throwable> onError) {
        return service.getVersions()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess(s -> version.set(s[0]))
                .doOnEvent((strings, throwable) -> onComplete.run())
                .doOnError(onError)
                .subscribe();
    }

    public String getVersion() {
        return version.get();
    }


    public Disposable getChampionList(@NonNull Consumer<? super List<Champion>> subscriber, @NonNull Consumer<Throwable> onError) {
        return service.getChampions(getVersion(), getLocal())
                .subscribeOn(Schedulers.io())
                .flatMapObservable(Observable::fromIterable)
                .filter(champion -> champion != null)
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .doOnError(onError)
                .subscribe(subscriber);
    }

    public Disposable verifyImages(List<Champion> champions, ProgressCallback callback, Consumer<List<ImageDescriptor>> consumer, @NonNull Consumer<Throwable> onError) {
        final List<ImageDescriptor> newImages = getNewImages(champions, getCompressionMethod());
        final AtomicInteger downloadCount = new AtomicInteger();
        final int total = newImages.size();
        return Observable.fromIterable(newImages)
                .subscribeOn(Schedulers.io())
                .flatMap(item -> Observable.just(item)
                        .subscribeOn(Schedulers.io())
                        .map(ImageDescriptor::verifySavedFile), MAX_CONCURRENCY)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext(p -> callback.onProgressUpdate(VERIFY_SUCCESS, downloadCount.incrementAndGet(), total))
                .doOnComplete(callback::finishExecution)
                .filter(ImageDescriptor::isInValid)
                .distinct()
                .toList()
                .doOnSuccess(consumer)
                .doOnError(onError)
                .subscribe();
    }

    public Disposable downloadAllImages(List<ImageDescriptor> champions, @NonNull ProgressCallback callback, @NonNull Action onComplete, @NonNull Consumer<Throwable> onError) {
        final AtomicInteger downloadCount = new AtomicInteger();
        final int total = champions.size();
        final Bitmap.CompressFormat compressFormat = getCompressionMethod();
        final int quality = getQuality();
        return Observable.fromIterable(champions)
                .subscribeOn(Schedulers.io())
                .flatMap(item -> Observable.just(item)
                        .subscribeOn(Schedulers.io())
                        .map(imDesc -> imDesc.verifyDownload(this, compressFormat, quality)), MAX_CONCURRENCY)

                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext(p -> callback.onDownloadSuccess(downloadCount.incrementAndGet(), total))
                .doOnComplete(callback::finishExecution)
                .doOnComplete(onComplete)
                .doOnError(onError)
                .subscribe();
    }

    private List<ImageDescriptor> getNewImages(List<Champion> champions, Bitmap.CompressFormat format) {
        List<ImageDescriptor> list = new ArrayList<>();
        for (Champion champion : champions) {
            for (ImageType type : ImageType.values()) {
                try {
                    File file = getChampionFile(champion.getId(), type, format);
                    list.add(new ImageDescriptor(champion.getId(), type, file));
                } catch (IOException e) {
                    Log.e(TAG, "getNewImages: ", e);
                }
            }
        }
        return list;
    }

    public void deleteChampionImages() throws IOException {
        FileStorage storage = FileStorage.getInstance();
        File dir = storage.getChampionImageDir();
        storage.deleteRecursive(dir);
    }

    @NonNull
    private File getChampionFile(@NonNull String champion, @NonNull ImageType type, @NonNull Bitmap.CompressFormat format) throws IOException {
        FileStorage storage = FileStorage.getInstance();
        return new File(storage.getChampionImageDir(), String.format("%s_%s.%s", champion, type.name().toLowerCase(), format.name().toLowerCase()));
    }

    public Bitmap getChampionBitmap(@NonNull Champion champion, @NonNull ImageType type) throws IOException {
        File file = getChampionFile(champion.getId(), type, getCompressionMethod());
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inMutable = false;
        options.inPreferredConfig = Bitmap.Config.ARGB_4444;
        return BitmapFactory.decodeFile(file.getPath(), options);
    }

    protected void saveBitmap(@NonNull final File file, final Bitmap bitmap, Bitmap.CompressFormat compressFormat, int quality) throws IOException {
        if (bitmap != null && (file.exists() || file.createNewFile())) {
            try (FileOutputStream outputStream = new FileOutputStream(file)) {
                bitmap.compress(compressFormat, quality, outputStream);
            }
        }
    }

    @NonNull
    protected Maybe<ResponseBody> getChampionCall(String champion, ImageType type, int skinId) {
        switch (type) {
            case SQUARE:
                return service.getSquareImage(getVersion(), champion);
            case SPLASH:
            default:
                return service.getSplashImage(champion, skinId);
        }
    }

    private DDragonService createService() {
        Gson gson = new GsonBuilder()
                .registerTypeAdapter(new TypeToken<List<Champion>>() {}.getType(), getChampionsDeserializer())
                .create();
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
                .build();
        return retrofit.create(DDragonService.class);
    }

    private JsonDeserializer<List<Champion>> getChampionsDeserializer() {
        return (json, typeOfT, context) -> {
            JsonObject object = json.getAsJsonObject();
            object = object.getAsJsonObject("data");
            List<Champion> list = new ArrayList<>();
            Set<Map.Entry<String, JsonElement>> entries = object.entrySet();
            for (Map.Entry<String, JsonElement> entry : entries) {
                Champion champion = context.deserialize(entry.getValue(), Champion.class);
                if (champion != null) {
                    list.add(champion);
                }
            }
            return list;
        };
    }

    private int getQuality() {
        return preferences.getInt("pref_image_quality", resources.getInteger(R.integer.pref_image_quality_default));
    }

    private String getLocal() {
        return preferences.getString("pref_language", resources.getString(R.string.pref_language_default));
    }

    private Bitmap.CompressFormat getCompressionMethod() {
        return Bitmap.CompressFormat.valueOf(preferences.getString("pref_image_type", resources.getString(R.string.pref_title_image_type_default)));
    }

}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/ddragon/DDragon.kt
package com.spiderbiggen.randomchampionselector.ddragon

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.preference.PreferenceManager
import android.provider.MediaStore.getVersion
import android.util.Log
import android.widget.ImageView
import com.spiderbiggen.randomchampionselector.R
import com.spiderbiggen.randomchampionselector.model.Champion
import com.spiderbiggen.randomchampionselector.model.ImageType
import com.spiderbiggen.randomchampionselector.storage.BitmapCache
import com.spiderbiggen.randomchampionselector.storage.file.FileStorage
import com.spiderbiggen.randomchampionselector.util.async.ProgressCallback
import com.spiderbiggen.randomchampionselector.util.async.ProgressCallback.Progress.VERIFY_SUCCESS
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

/**
 * Created on 13-3-2018.
 *
 * @author Stefan Breetveld
 */
object DDragon {

    private val TAG = DDragon::class.java.simpleName
    private const val BASE_URL = "http://ddragon.leagueoflegends.com"
    // Default version if versions endpoint fails
    private const val DEFAULT_VERSION = "8.4.1"
    private const val MAX_CONCURRENCY = 8
    private val version = AtomicReference(DEFAULT_VERSION)

    private val service: DDragonService
    private var preferences: SharedPreferences? = null
    private var resources: Resources? = null

    private val championsDeserializer: Converter.Factory
        get() = JsonConverterFactory()

    private val quality: Int
        get() = preferences?.getInt("pref_image_quality", resources?.getInteger(R.integer.pref_image_quality_default)
                ?: 85) ?: 85

    private val local: String
        get() = preferences?.getString("pref_language", resources?.getString(R.string.pref_language_default))
                ?: "en_US"

    private val compressionMethod: Bitmap.CompressFormat
        get() = Bitmap.CompressFormat.valueOf(preferences?.getString("pref_image_type", resources?.getString(R.string.pref_title_image_type_default))
                ?: "WEBP")

    init {
        this.service = createService()
    }

    /**
     * Sets preferences.
     *
     * @param preferences the new value of preferences
     */
    fun setPreferences(preferences: SharedPreferences) {
        this.preferences = preferences
    }

    /**
     * Sets resources.
     *
     * @param resources the new value of resources
     */
    fun setResources(resources: Resources) {
        this.resources = resources
    }

    fun useContext(context: Context) {
        setPreferences(PreferenceManager.getDefaultSharedPreferences(context.applicationContext))
        setResources(context.applicationContext.resources)
    }

    fun updateVersion(onComplete: Action, onError: Consumer<Throwable>): Disposable {
        return service.versions
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess { version.set(it[0]) }
                .doOnEvent { _, _ -> onComplete.run() }
                .doOnError(onError)
                .subscribe()
    }

    fun getChampionList(subscriber: Consumer<in List<Champion>>,
                        onError: Consumer<Throwable>): Disposable {
        return service.getChampions(version.get(), local)
                .subscribeOn(Schedulers.io())
                .flatMapObservable<Champion> { Observable.fromIterable(it) }
//                .filter { it != null }
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .doOnError(onError)
                .subscribe(subscriber)
    }

    fun verifyImages(champions: List<Champion>, callback: ProgressCallback, consumer: Consumer<List<ImageDescriptor>>, onError: Consumer<Throwable>): Disposable {
        val newImages = getNewImages(champions, compressionMethod)
        val downloadCount = AtomicInteger()
        val total = newImages.size
        return Observable.fromIterable(newImages)
                .subscribeOn(Schedulers.io())
                .flatMap({
                    Observable.just(it)
                            .subscribeOn(Schedulers.io())
                            .map<ImageDescriptor> { it.verifySavedFile() }
                }, MAX_CONCURRENCY)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext { callback.onProgressUpdate(VERIFY_SUCCESS, downloadCount.incrementAndGet(), total) }
                .doOnComplete { callback.finishExecution() }
                .filter { it.invalid }
                .distinct()
                .toList()
                .doOnSuccess(consumer)
                .doOnError(onError)
                .subscribe()
    }

    fun downloadAllImages(champions: List<ImageDescriptor>, callback: ProgressCallback,
                          onComplete: Action, onError: Consumer<Throwable>): Disposable {
        val downloadCount = AtomicInteger()
        val total = champions.size
        val compressFormat = compressionMethod
        val quality = quality
        return Observable.fromIterable(champions)
                .subscribeOn(Schedulers.io())
                .flatMap({
                    Observable.just(it)
                            .subscribeOn(Schedulers.io())
                            .map { it.verifyDownload(compressFormat, quality) }
                }, MAX_CONCURRENCY)

                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext { callback.onDownloadSuccess(downloadCount.incrementAndGet(), total) }
                .doOnComplete { callback.finishExecution() }
                .doOnComplete(onComplete)
                .doOnError(onError)
                .subscribe()
    }

    private fun getNewImages(champions: List<Champion>, format: Bitmap.CompressFormat): List<ImageDescriptor> {
        val list = ArrayList<ImageDescriptor>()
        for (champion in champions) {
            for (type in ImageType.values()) {
                try {
                    val file = getChampionFile(champion.id, type, format)
                    list.add(ImageDescriptor(champion.id, type, file))
                } catch (e: IOException) {
                    Log.e(TAG, "getNewImages: ", e)
                }

            }
        }
        return list
    }

    @Throws(IOException::class)
    fun deleteChampionImages() {
        val storage = FileStorage
        val dir = storage.championImageDir
        storage.deleteRecursive(dir)
    }

    @Throws(IOException::class)
    private fun getChampionFile(champion: String, type: ImageType, format: Bitmap.CompressFormat): File =
            File(FileStorage.championImageDir, "${champion}_${type.name.toLowerCase()}.${format.name.toLowerCase()}")

    @Deprecated("Should probably load this asynchronously.", replaceWith = ReplaceWith("getChampionBitmapFromCache"), level = DeprecationLevel.WARNING)
    @Throws(IOException::class)
    fun getChampionBitmap(champion: Champion, type: ImageType): Bitmap? {
        val file = getChampionFile(champion.id, type, compressionMethod)
        val options = BitmapFactory.Options()
        options.inMutable = false
        options.inPreferredConfig = Bitmap.Config.ARGB_4444
        return BitmapFactory.decodeFile(file.path, options)
    }

    fun getChampionBitmapFromCache(champion: Champion, type: ImageType, imageView: ImageView) {
        val file = getChampionFile(champion.id, type, compressionMethod)
//        val options = BitmapFactory.Options()
//        options.inMutable = false
//        options.inPreferredConfig = Bitmap.Config.ARGB_4444
        return BitmapCache.loadBitmap("${champion.key}_${type.name.toLowerCase()}", file, imageView)
    }

    @Throws(IOException::class)
    fun saveBitmap(file: File, bitmap: Bitmap?, compressFormat: Bitmap.CompressFormat, quality: Int) {
        if (bitmap != null && (file.exists() || file.createNewFile())) {
            FileOutputStream(file).use { bitmap.compress(compressFormat, quality, it) }
        }
    }

    fun getChampionCall(champion: String, type: ImageType, skinId: Int): Maybe<ResponseBody> {
        return when (type) {
            ImageType.SQUARE -> service.getSquareImage(version.get(), champion)
            ImageType.SPLASH -> service.getSplashImage(champion, skinId)
        }
    }

    private fun createService(): DDragonService {
        val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(championsDeserializer)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
                .build()
        return retrofit.create(DDragonService::class.java)
    }

}
